app/models/piece.rb

```diff
@@ -20,46 +29,104 @@ def move_to!(new_x, new_y)
     end
     # Finally, it should call update_attributes on the piece and change the piece's x/y position.
     # http://apidock.com/rails/ActiveRecord/Base/update_attributes
-    raise ArgumentError, "Can't move piece" unless update_attributes(x_coordinate: new_x, y_coordinate: new_y)
+    raise ArgumentError, "Can't move piece" unless update_attributes(x_coordinate: x, y_coordinate: y)
   end
 
-  def on_board?(destination_x, destination_y)
-    return false if destination_x < 0
-    return false if destination_x > 7
-    return false if destination_y < 0
-    return false if destination_y > 7
-    true
+  def obstructed?(x, y)
+    # only check if obstructed in the direction you are moving
+    return true if vertical?(x, y) && obstructed_vertically?(y)
+    return true if horizontal?(x, y) && obstructed_horizontally?(x)
+    return true if diagonal?(x, y) && obstructed_diagonally?(x, y)
+    false
+  end
+
+  def square_is_occupied(x, y)
+    # returns the square is occupied
+    game.pieces.where(x_coordinate: x, y_coordinate: y, captured: false).exists?
+  end
+
+  # checks for horizontal obstruction
+  def obstructed_horizontally?(x)
+    x_distance = (x_coordinate - x).abs
+
+    # if moving right, x_dir is positive
+    x_dir = x > x_coordinate ? 1 : -1
+
+    # this checks if there are any steps in between the coordinate and destination
+    return false if x_distance < 2
+    # distance -1 means we go one less than the distance because we don't count the end distance
+    (1..(x_distance - 1)).each do |step| # number of steps horizontally
+      x_steps = step * x_dir
+      y_steps = step * 0
+      x_square = x_coordinate + x_steps
+      y_square = y_coordinate + y_steps
+
+      return true if square_is_occupied(x_square, y_square)
+    end
+    false
```

Comment:
I know these 3 obstructed functions could be refactored more, but my brain can't handle anymore right now. :(